docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,277 @@
+    2. To update custom resource spec, K8s user will have to check resource status because status will have most updated
+     value for some of the fields.
+
+
+
+### Solution 3 (Preferred Solution)
+
+In this solution, If a default value is added or existing value is updated as part of Create/Update API output, ACK service controller will update the spec/status of the custom resource accordingly. These updates will be persisted in custom resource’s Annotation to refer in future reconciliations.
+
+
+1. *Consistency with K8s conventions* ✅
+    1. In K8s convention, the spec fields are not duplicated into status fields.
+    2. However we will be updating the desired state of custom resource from ACK service controller, which is not really the recommended k8s way.
+2. *Consistency across ACK service controllers* ✅
+    1. All ACK service controllers will be consistent. Spec fields in each custom resource will represent the desired state of the cluster.
+3. *GitOps compatibility* ✅
+    1. There should be no complexity in implementing GitOps for ACK custom resources. 
+    2. Approach mentioned in low level design below should work.
+4. *User experience for custom resource lifecycle* ✅
+    1. Applying updates to custom resource will not be confusing for the user. Users will update the spec and service controller will create Update request based on new spec.
+
+
+> I personally prefer “Solution 3” provided it gives a consistent GitOps behavior.
+
+> The biggest selling point for “Solution 3” is that it is consistent with the K8s convention where desired state is 
+> defined by spec of the custom resource and also custom resource spec in “etcd” is consistent with AWS resource.
+
+> To make the Update resource user experience better, In addition to the Status’s condition, we will also add annotation
+> capturing the changes which were made to desired spec. [More details in Low Level Design]
+
+
+## High Level Design
+
+![High Level Design](./images/late_initialization.png)
+[source](./images/late_initialization.drawio)
+
+
+## POC
+
+I did a small POC to determine how GitOps(Fluxcd) behaves when desired spec is updated by the controller.
+
+Setup:
+
+* A basic “Guestbook” [operator](https://github.com/vijtrip2/guestbook/blob/main/controllers/guestbook_controller.go) from
+ kubebuilder getting started guide. 
+* “Guestbook” [manifest](https://github.com/vijtrip2/flux-get-started/blob/master/guestbook/guestbook-sample.yaml) being
+ deployed using fluxcd. 
+* Controller adds a defaultValue in spec on each run, updates an existing value in spec and updates the annotation on each
+ reconcile loop.
+
+Observations:
+
+* During the start of each reconciliation, the spec and status are reset to values present in the git repository.
+* Annotations are not reset, if a value is updated in annotations, it is present for future reconciliations.
+* Changes made after reconciliation are not reflected back into the git repository. “Fluxcd” does not support this. 
```

Comment:
What do you mean by `reflected back into the git repository`? A new commit modifying the GitOps repository? I'd sure hope no K8s CD system modifies the repo haha

Reply:
Yeah, that's what I meant.
Fluxcd does update the git repo back but only if you make changes to k8s object using flux api.